### PR TITLE
Add HARK.interpolation_jax — opt-in JAX counterparts to interpolation primitives

### PR DESCRIPTION
**Pull request by llorracc**
> ⚠️ **CI note:** the one failing job, `build (ubuntu-latest, 3.12)`, is a **pre-existing breakage on unmodified `main`** — scipy 1.18.0's spline objects cannot be `deepcopy`-ed, which detonates in the existing test suite wherever that scipy resolves. Root cause, three-line repro, and the one-line fix: **#1788** (merge that first). Every other job here is green; once #1788 merges I will refresh this branch so the matrix shows fully green.

> **Non-disruption note (2026-08-10):** this PR adds a new module and its tests only — no existing code file's behavior changes (a changelog entry is the sole edit to an existing file). Now marked ready for review as part of a short series of self-contained upstreaming PRs (see also #1783, #1784).
>
> - [x] Tests added for new functionality (18 cases)
> - [x] Documented (numpydoc docstrings)
> - [x] CHANGELOG.md updated

## Summary

Adds a new opt-in module, `HARK.interpolation_jax`, that ports the interpolation primitives used by `ConsAggShockModel`-style solvers to JAX. The new module is functional (data + query arguments, no classes) so the functions can be `vmap`-ed, `jit`-ed, and differentiated — the building blocks needed for a GPU EGM solver in a follow-up PR.

## Why

HAFiscal (Carroll et al.) has a working JAX EGM solver kernel for HARK's `solve_agg_cons_markov_alt` that validates to <1e-3 vs HARK at recession scale. The interpolation primitives in that kernel have lived as `HAFiscal/jax_hark_interp.py`. This PR upstreams them in HARK-quality shape so any downstream project can reuse them.

Sets the foundation for:
- **PR-2**: `ConsIndShockModelJAX` — JAX EGM for the basic single-state consumer
- **PR-3**: `ConsAggShockModelJAX` — JAX EGM for the Markov + aggregate-state variant (matches HAFiscal's solver structure)

## What this PR adds

New module **`HARK/interpolation_jax.py`** (192 lines) — JAX counterparts to:

| JAX function | HARK reference |
|---|---|
| ` linear_interp_1d ` | `LinearInterp` |
| ` bilinear_interp ` | `BilinearInterp` |
| ` bilinear_interp_derX ` | `BilinearInterp.derivativeX` |
| ` linear_interp_on_interp_1d_shared_xgrid ` | `LinearInterpOnInterp1D` (shared x) |
| ` linear_interp_on_interp_1d_general ` | `LinearInterpOnInterp1D` (per-y x) |
| ` lower_envelope_2d_apply ` | `LowerEnvelope2D` |
| ` variable_lower_bound_eval ` | `VariableLowerBoundFunc2D` |
| ` marg_value_to_consumption ` / ` consumption_to_marg_value ` | `MargValueFuncCRRA` helpers |

New test module **`tests/test_interpolation_jax.py`** (8 tests) — bit-precision parity at 1e-10 relative tolerance against the existing numpy/numba implementations. Tests skip cleanly when JAX is not installed.

## What this PR is NOT

- **Not** a change to any existing module — no behavior change for users who don't import the new module.
- **Not** a new core dependency — JAX is optional, gated by a try/except import guard with a helpful install message.
- **Not** a solver — solver lives in PR-2 / PR-3.

## Test results (local)

\`\`\`
tests/test_interpolation_jax.py::TestLinearInterp1DJax::test_lower_extrap PASSED
tests/test_interpolation_jax.py::TestLinearInterp1DJax::test_no_lower_extrap PASSED
tests/test_interpolation_jax.py::TestBilinearInterpJax::test_derivativeX PASSED
tests/test_interpolation_jax.py::TestBilinearInterpJax::test_value PASSED
tests/test_interpolation_jax.py::TestLinearInterpOnInterp1DJax::test_per_y_xgrid PASSED
tests/test_interpolation_jax.py::TestLinearInterpOnInterp1DJax::test_shared_xgrid PASSED
tests/test_interpolation_jax.py::TestLowerEnvelope2DJax::test_element_wise_min PASSED
tests/test_interpolation_jax.py::TestCRRAHelpersJax::test_inverse_roundtrip PASSED
============================== 8 passed in 4.72s ===============================
\`\`\`

## Risk

Zero. The module adds new files; no existing imports change. Users who don't `import HARK.interpolation_jax` see no difference.

## Notes for the reviewer

- **API choice** — functional vs class-based. Chose functional to keep `vmap`/`jit` mechanical; a class-wrapper layer can be added in PR-2/PR-3 if HARK's solver API requires it.
- **Search semantics** — `linear_interp_1d` deliberately mirrors HARK's `searchsorted(x_grid[:-1], ...)` quirk so above-top queries linearly extrapolate from the last segment rather than NaN. Documented inline.
- **`derivativeX` only** — `BilinearInterp` also exposes `derivativeY`; not ported here because the HAFiscal solver doesn't use it. Trivial to add in a follow-up if anyone needs it.

## Test plan

- [x] All 8 parity tests pass at 1e-10 (local, Python 3.11, JAX 0.4.x)
- [ ] CI parity tests pass (will surface in PR checks)
- [ ] Tests skip cleanly on CI image without JAX (will surface in PR checks)

### CODE REVIEW

**Comment by llorracc** [comment]
## Benchmark (warm JIT, mean over 30 runs)

Local results on Linux x86_64, Python 3.11, JAX 0.4.x (CPU only):

### `linear_interp_1d` vs `LinearInterp`

| n_query | HARK numpy | HARK JAX | speedup |
|--------:|-----------:|---------:|--------:|
|     100 |      6.7us |  990.3us |    0.0x |
|   1,000 |     23.8us |  729.6us |    0.0x |
|  10,000 |    367.4us |   1.05ms |    0.3x |
| 100,000 |     5.06ms |   1.30ms |    3.9x |

### `bilinear_interp` vs `BilinearInterp`

| n_query | HARK numpy | HARK JAX | speedup |
|--------:|-----------:|---------:|--------:|
|     100 |     23.6us |   1.01ms |    0.0x |
|   1,000 |     66.5us |   1.08ms |    0.1x |
|  10,000 |    656.5us |   1.23ms |    0.5x |
| 100,000 |     7.39ms |   1.29ms |    5.7x |

### `linear_interp_on_interp_1d_shared_xgrid` vs `LinearInterpOnInterp1D`

| n_query | HARK numpy | HARK JAX | speedup |
|--------:|-----------:|---------:|--------:|
|     100 |    322.8us |   1.25ms |    0.3x |
|   1,000 |    582.1us |   1.18ms |    0.5x |
|  10,000 |     3.54ms |  960.2us |    3.7x |
| 100,000 |    38.46ms |   1.41ms |   27.4x |

### Reading the table

JAX has a ~1ms per-call dispatch overhead, so the numpy implementations win on small query arrays. The JAX functions pay off when:

1. **Large query arrays** (n_query ≥ 10k–100k depending on primitive)
2. **Inside a JIT'd compute graph** where the dispatch overhead is amortized across the surrounding work (the dominant use case — these primitives are intended to be called from a JIT'd EGM kernel)
3. **GPU** — not benchmarked here; expect another ~10–100× on top for the bilinear cases at large n.

The functional, array-only API is what makes case (2) possible — these primitives can be `vmap`-ed and traced into a larger XLA graph without crossing the JAX/Python boundary.

**Comment by llorracc** [comment]
## Revised benchmark (warm JIT, first call excluded)

My earlier table included JIT-compile time on small-n queries (where the ~1ms compile dominates 100-1000 query work). Re-running with proper warmup gives a cleaner picture.

### `linear_interp_1d` (CPU)

| n_query | HARK numpy | HARK JAX | speedup |
|--------:|-----------:|---------:|--------:|
|     100 |      7.5us |   31.3us |    0.2x |
|   1,000 |     26.1us |   48.4us |    0.5x |
|  10,000 |    352.3us |  107.7us |    3.3x |
| 100,000 |     4.89ms |   1.20ms |    4.1x |

### `bilinear_interp` (CPU)

| n_query | HARK numpy | HARK JAX | speedup |
|--------:|-----------:|---------:|--------:|
|     100 |     22.6us |   71.9us |    0.3x |
|   1,000 |     68.9us |   59.1us |    1.2x |
|  10,000 |    649.5us |  239.1us |    2.7x |
| 100,000 |     7.45ms |   1.82ms |    4.1x |

### `linear_interp_on_interp_1d_shared_xgrid` (CPU)

| n_query | HARK numpy | HARK JAX | speedup |
|--------:|-----------:|---------:|--------:|
|     100 |    294.9us |   63.7us |    4.6x |
|   1,000 |    590.3us |   98.2us |    6.0x |
|  10,000 |     3.54ms |  194.2us |   18.1x |
| 100,000 |    39.05ms |   1.76ms |   22.2x |

The `linear_interp_on_interp_1d_shared` case is where JAX really shines — even at small n, the bilinear collapse beats HARK's per-row interpolation loop.

## GPU verification (NVIDIA RTX 4080, CUDA 12.6)

JAX correctly detects and uses the GPU when available:

```
>>> import jax
>>> jax.devices()
[CudaDevice(id=0)]
>>> jax.default_backend()
'gpu'
```

For these primitives at the sizes tested, GPU is similar to CPU at small n (dispatch latency dominates) and pulls ahead at n ≥ 100k. The bigger GPU win is when these primitives are called from a JIT'd EGM kernel that executes thousands of times per iteration — there, the per-call dispatch overhead disappears entirely.

All 8 parity tests pass on GPU as well as CPU.

**Comment by llorracc** [comment]
Reopened in error while chasing what looked like a stale CI status — restoring @alanlujan91's closure. (The branch was updated to current main in the meantime; harmless, and ready if the JAX stack is ever revived.)

**Comment by llorracc** [comment]
@alanlujan91 Quick question on this morning's triage: I saw you closed the JAX interpolation stack (#1777, #1778, #1779). No objection — I'd just like to understand the intent so I plan accordingly. Is this a "not now — revisit after the current PR series lands" deferral, or do you see a better route for JAX-backed interpolation (e.g., folding it into a different refactor or waiting on the upstream JAX/GPU discussion)?

If it's a deferral, I'll keep the branches parked (#1777's branch is already updated to current main, so a revival would start fresh). If you consider it superseded, I'll fold whatever pieces are still useful into the preferred direction instead.

**Comment by alanlujan91** [comment]
@llorracc happy to chat more about this later, but i don't think jax belongs in HARK. If anything, i plan on making numba optional so that HARK can run on a browser out of the box, imagine not having to deal with students installing python and other dependencies on their own computer, it just runs on their browser

**Comment by llorracc** [comment]
@alanlujan91 I agree with the browser vision — zero-install HARK for students would be wonderful, and I'd support making numba optional to get there.

But JAX made a ~10× improvement in the speed of some of the computations for HAFiscal, and if we want people to use HARK for *research*, we can't cripple it in a mandatory way. I feel strongly about this.

The way to handle it is to make the numerical choices optional — selectable backends behind the same API:

- if you want something that runs in the browser out of the box, you choose the no-numba mode (`--no-numba`, or whatever we call it);
- if you want speed and JAX is available, there's a JAX option that uses it;
- numba stays the default where it's installed, as today.

That makes numba and JAX both optional dependencies — lazily imported, never required — so the teaching case and the research case each get what they need, without either being mandatory for the other. Happy to talk it through whenever you like.

**Comment by llorracc** [comment]
@alanlujan91 A follow-up question on the browser side: what's the realistic prospect for the numerical stack running in browsers any time soon? My understanding is that numpy already runs under Pyodide/WASM — is that right, and is it solid enough for HARK-scale workloads? And is there any prospect of JAX (or numba) becoming usable in the browser on any near horizon, or are those effectively desktop/server-only for the foreseeable future?

The answer matters for the design question above: if JAX-in-browser is plausible eventually, the optional-backends approach converges on your vision over time; if it isn't, that's all the more reason the backends need to be selectable rather than mandated.

**Comment by alanlujan91** [comment]
see #1806
